go/vt/vtgate/vindexes/cfc.go

```diff
@@ -0,0 +1,326 @@
+// it easier to reason the fanout but the data distribution depends on the key
+// itself; while using the hash on components takes care of the randomness of
+// the data distribution.
+//
+// Since the vindex is on a concatenated key, the offsets into the key are the
+// only way to mark its components. Thus it implicitly requires each component
+// to have a fixed width, except the last one. It's especially true when hash
+// is defined. Because the hash is calculated component by component, only the
+// prefix that aligns with the component boundary can be used to compute the
+// key range. Although the misaligned part doesn't participate the key range
+// calculation, the SQL executed on each shard uses the unchanged prefix; thus
+// the behavior is exactly same as other vindex's but just more efficient in
+// controlling the fanout.
+//
+// The expected format of the vindex definition is
+//
+// "vindexes": {
+//   "cfc_md5": {
+//     "type": "cfc",
+//     "params": {
+//       "hash": "md5",
+//       "offsets": "[2,4]"
+//     }
+//   }
+// }
+//
+// 'offsets' only makes sense when hash is used. Offsets should be a sorted
+// list of positive ints, each of which denotes the byte offset (from the
+// beginning of key) of each component's boundary in the concatenated key.
+// Specifically, offsets[0] is the byte offset of the first component,
+// offsets[1] is the byte offset of the second component, etc.
+type CFC struct {
+	name         string
+	hash         func([]byte) []byte
+	offsets      []int
+	prefixVindex SingleColumn
+}
+
+// NewCFC creates a new CFC vindex
+func NewCFC(name string, params map[string]string) (Vindex, error) {
+	// CFC is used in all compare expressions other than 'LIKE'.
+	ss := &CFC{
+		name: name,
+	}
+	// prefixCFC is only used in 'LIKE' compare expressions.
+	ss.prefixVindex = &prefixCFC{name: name, cfc: ss}
+
+	if params == nil {
+		return ss, nil
+	}
+
+	switch h := params["hash"]; h {
+	case "":
+		return ss, nil
+	case "md5":
+		ss.hash = md5hash
+	case "xxhash64":
+		ss.hash = xxhash64
+	default:
+		return nil, fmt.Errorf("invalid hash %s to CFC vindex %s", h, name)
+	}
+
+	var offsets []int
+	if p := params["offsets"]; p == "" {
+		return nil, fmt.Errorf("CFC vindex requires offsets when hash is defined")
+	} else if err := json.Unmarshal([]byte(p), &offsets); err != nil || !validOffsets(offsets) {
+		return nil, fmt.Errorf("invalid offsets %s to CFC vindex %s. expected sorted positive ints in brackets", p, name)
+	}
+	// remove duplicates
+	prev := -1
+	for _, off := range offsets {
+		if off != prev {
+			ss.offsets = append(ss.offsets, off)
+		}
+		prev = off
+	}
+
+	return ss, nil
+}
+
+func validOffsets(offsets []int) bool {
+	if len(offsets) == 0 || !sort.IntsAreSorted(offsets) {
+		return false
+	}
+	if offsets[0] <= 0 {
+		return false
+	}
+	return true
+}
+
+func (vind *CFC) String() string {
+	return vind.name
+}
+
+// Cost returns the cost as 1.
+func (vind *CFC) Cost() int {
+	return 1
```

Comment:
Did you consider other vindex costs when you arrived at this value?

Reply:
I'm not so familiar the numeric system to infer the vindex cost. Do you have suggestions for how to calculate the cost ?

Reply:
can you change the cost to sum of cost of hash function involved for each column

Reply:
which is no. of column * cost of hash function

Reply:
when `offsets` parameter is set, we know the number of columns but if it isn't set, we have no idea how many columns are involved. What do you suggest in this case ?